library/alloc/tests/boxed.rs

```diff
@@ -57,3 +58,71 @@ fn box_deref_lval() {
     x.set(1000);
     assert_eq!(x.get(), 1000);
 }
+
+pub struct ConstAllocator;
+
+unsafe impl const Allocator for ConstAllocator {
+    fn allocate(&self, layout: Layout) -> Result<NonNull<[u8]>, AllocError> {
+        unsafe {
+            let ptr = core::intrinsics::const_allocate(layout.size(), layout.align());
+            Ok(NonNull::new_unchecked(ptr as *mut [u8; 0] as *mut [u8]))
+        }
+    }
+
+    unsafe fn deallocate(&self, _ptr: NonNull<u8>, _layout: Layout) {
+        /* do nothing */
+    }
+
+    fn allocate_zeroed(&self, layout: Layout) -> Result<NonNull<[u8]>, AllocError> {
+        self.allocate(layout)
+    }
+
+    unsafe fn grow(
+        &self,
+        _ptr: NonNull<u8>,
+        _old_layout: Layout,
+        _new_layout: Layout,
+    ) -> Result<NonNull<[u8]>, AllocError> {
+        unimplemented!()
+    }
+
+    unsafe fn grow_zeroed(
+        &self,
+        _ptr: NonNull<u8>,
+        _old_layout: Layout,
+        _new_layout: Layout,
+    ) -> Result<NonNull<[u8]>, AllocError> {
+        unimplemented!()
+    }
+
+    unsafe fn shrink(
+        &self,
+        _ptr: NonNull<u8>,
+        _old_layout: Layout,
+        _new_layout: Layout,
+    ) -> Result<NonNull<[u8]>, AllocError> {
+        unimplemented!()
```

Comment:
Please implement these.

Reply:
This is a type for testing, so I omitted some implementations. Should I still implement these?

Reply:
It is fine to leave them unimplemented for now, but I think they should be implemented before merging. You may be able to copy the default implementation of these methods.

Reply:
I have implemented these, but the methods aren't called in the testcases, we can't check if they are implemented correctly. Also we may need to implement `const_deallocate` to the compiler.

Reply:
Const deallocation is simply leaking the thing. The mir interpreter cleans up all dangling allocations

Reply:
> Also we may need to implement const_deallocate to the compiler.

I just saw you open a PR to add this, let me emphasize: This PR cannot contain any use of that intrinsic. The nature of `ConstAlloc`/`CtfeAlloc` is to leak when dropping, which means you can't realistically use it in this test allocator impl.

Reply:
>  The nature of ConstAlloc/CtfeAlloc is to leak when dropping

And why is that?
I don't think this should be the case, at least for allocations that are created and then deallocated within the same const-evalaution. (Of course, *at runtime* deallocation is a NOP, since static memory cannot be deallocated and it is shared among all the copies of the constant anyway.)

Reply:
Well, looks like `const_eval_select` can resolve this problem faily easily.

Reply:
I don't think that is enough -- if one `const` works on something that was allocated *in a different `const`*, deallocation should still be a NOP.